packages/flutter/lib/src/cupertino/segmented_control.dart

```diff
@@ -0,0 +1,568 @@
+  ///           setState(() {
+  ///             currentValue = newValue;
+  ///           });
+  ///         },
+  ///         groupValue: currentValue,
+  ///       ),
+  ///     );
+  ///   }
+  /// }
+  /// ```
+  final ValueChanged<T> onValueChanged;
+
+  @override
+  _SegmentedControlState<T> createState() => _SegmentedControlState<T>();
+}
+
+class _SegmentedControlState<T> extends State<SegmentedControl<T>> {
+  T _selectedValue;
+  T _pressedValue;
+  final List<Widget> _gestureChildren = <Widget>[];
+
+  @override
+  void initState() {
+    super.initState();
+    for (T currentKey in widget.children.keys) {
+      if (currentKey == widget.groupValue) {
+        _selectedValue = currentKey;
+      }
+    }
+  }
+
+  void _onTapDown(T currentKey) {
+    setState(() {
+      _pressedValue = currentKey;
+    });
+  }
+
+  void _onTapUp(TapUpDetails event) {
+    setState(() {
+      _pressedValue = null;
+    });
+  }
+
+  void _onTapCancel() {
+    setState(() {
+      _pressedValue = null;
+    });
+  }
+
+  void _onTap(T currentKey) {
+    setState(() {
+      if (_selectedValue != currentKey) {
+        widget.onValueChanged(currentKey);
+      }
+    });
+  }
+
+  @override
+  void didUpdateWidget(SegmentedControl<T> oldWidget) {
+    super.didUpdateWidget(oldWidget);
+    for (T currentKey in widget.children.keys) {
+      if (currentKey == widget.groupValue) {
+        _selectedValue = currentKey;
+      }
+    }
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    _gestureChildren.clear();
+    for (T currentKey in widget.children.keys) {
+      final TextStyle textStyle = DefaultTextStyle.of(context).style.copyWith(
+        color: (_selectedValue == currentKey) ?
+        CupertinoColors.white : CupertinoColors.activeBlue,
+      );
+      final IconThemeData iconTheme = IconThemeData(
+        color: (_selectedValue == currentKey) ?
+        CupertinoColors.white : CupertinoColors.activeBlue,
+      );
+
+      Widget child = widget.children[currentKey];
+      child = GestureDetector(
+        onTapDown: (TapDownDetails event) {
+          _onTapDown(currentKey);
+        },
+        onTapUp: _onTapUp,
+        onTapCancel: _onTapCancel,
+        onTap: () {
+          _onTap(currentKey);
+        },
+        child: new IconTheme(
+            data: iconTheme,
+            child: DefaultTextStyle(
+              style: textStyle,
+              overflow: TextOverflow.ellipsis,
+              child: Semantics(
+                inMutuallyExclusiveGroup: true,
+                selected: _selectedValue != null ? _selectedValue == currentKey : false,
+                child: child,
+              ),
+            )),
+      );
+      _gestureChildren.add(child);
+    }
+
+    final Widget box = new _SegmentedControlRenderWidget<T>(
+      children: _gestureChildren,
+      selectedValue: _selectedValue,
+      pressedValue: _pressedValue,
+      keys: widget.children.keys.toList(),
```

Comment:
You could just pass in the index here, rather than the whole list of keys. Since you're already iterating over the list, you could just track the index, and figure out the index for the selection and pressed value, then pass that in here.

Reply:
So if we did this, we would no longer need a selectedValue or pressedValue, right? Instead, we would have something like a pressedIndex and selectedIndex. We're having a little trouble because you can't iterate through a Map using an index, so it seems like using an int index would result in more processing (due to all these iterations through the keys) as opposed to just passing down the reference to the key itself. Let us know if we should come talk to you about this in person!

Reply:
I just mean for passing down to the render object. The render object doesn't have the map so it's not going to iterate through it; it just has a list and iterates through that, so indices are already what it deals with.